pyro/contrib/named.py

```diff
@@ -0,0 +1,245 @@
+"""
+Named Data Structures
+---------------------
+
+The ``pyro.contrib.named`` module is a thin syntactic layer on top of Pyro.  It
+allows Pyro models to be written to look like programs with operating on Python
+data structures like ``latent.x.sample_(...)``, rather than programs with
+string-labeled statements like ``x = pyro.sample("x", ...)``.
+
+This module provides three container data structures ``named.Object``,
+``named.List``, and ``named.Dict``. These data structures are intended to be
+nested in each other. Together they track the address of each piece of data
+in each data structure, so that this address can be used as a Pyro site. For
+example::
+
+    >>> state = named.Object("state")
+    >>> print(str(state))
+    state
+
+    >>> z = state.x.y.z  # z is just a placeholder.
+    >>> print(str(z))
+    state.x.y.z
+
+    >>> state.xs = named.List()  # Create a contained list.
+    >>> x0 = state.xs.add()
+    >>> print(str(x0))
+    state.xs[0]
+
+    >>> state.ys = named.Dict()
+    >>> foo = state.ys['foo']
+    >>> print(str(foo))
+    state.ys['foo']
+
+These addresses can now be used inside ``sample``, ``observe`` and ``param``
+statements. These named data structures even provide in-place methods that
+alias Pyro statements. For example::
+
+    >>> state = named.Object("state")
+    >>> mu = state.mu.param_(Variable(torch.zeros(1), requires_grad=True))
+    >>> sigma = state.sigma.param_(Variable(torch.ones(1), requires_grad=True))
+    >>> z = state.z.sample_(dist.normal, mu, sigma)
+    >>> state.x.observe_(dist.normal, z, mu, sigma)
+
+For deeper examples of how these can be used in model code, see the
+`Tree Data <https://github.com/uber/pyro/blob/dev/examples/contrib/named/tree_data.py>`_
+and
+`Mixture <https://github.com/uber/pyro/blob/dev/examples/contrib/named/mixture.py>`_
+examples.
+
+Authors: Fritz Obermeyer, Alexander Rush
```

Comment:
@srush I've added you as a coauthor since you suggested most of the ideas and you'll be adding more code in #550 .